examples/petsc/src/libceedsetup.c

```diff
@@ -202,54 +197,52 @@ PetscErrorCode SetupLibceedByDegree(DM dm, Ceed ceed, CeedInt degree,
 // -----------------------------------------------------------------------------
 // Setup libCEED level transfer operator objects
 // -----------------------------------------------------------------------------
-PetscErrorCode CeedLevelTransferSetup(Ceed ceed, CeedInt num_levels,
+PetscErrorCode CeedLevelTransferSetup(DM dm, Ceed ceed, CeedInt level,
+                                      CeedInt num_levels,
                                       CeedInt num_comp_u, CeedData *data,
-                                      CeedInt *level_degrees,
-                                      CeedQFunction qf_restrict, CeedQFunction qf_prolong) {
+                                      CeedInt *level_degrees, Vec fine_mult) {
+  int ierr;
   // Return early if num_levels=1
   if (num_levels == 1)
     PetscFunctionReturn(0);
 
-  // Set up each level
-  for (CeedInt i=1; i<num_levels; i++) {
-    // P coarse and P fine
-    CeedInt Pc = level_degrees[i-1] + 1;
-    CeedInt Pf = level_degrees[i] + 1;
-
-    // Restriction - Fine to corse
-    CeedBasis basis_c_to_f;
-    CeedOperator op_restrict;
-
-    // Basis
-    CeedBasisCreateTensorH1Lagrange(ceed, 3, num_comp_u, Pc, Pf,
-                                    CEED_GAUSS_LOBATTO, &basis_c_to_f);
-
-    // Create the restriction operator
-    CeedOperatorCreate(ceed, qf_restrict, CEED_QFUNCTION_NONE,
-                       CEED_QFUNCTION_NONE, &op_restrict);
-    CeedOperatorSetField(op_restrict, "input", data[i]->elem_restr_u,
-                         CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
-    CeedOperatorSetField(op_restrict, "output", data[i-1]->elem_restr_u,
-                         basis_c_to_f, CEED_VECTOR_ACTIVE);
-
-    // Save libCEED data required for level
-    data[i]->basis_c_to_f = basis_c_to_f;
-    data[i]->op_restrict = op_restrict;
-
-    // Interpolation - Corse to fine
-    CeedOperator op_prolong;
-
-    // Create the prolongation operator
-    CeedOperatorCreate(ceed, qf_prolong, CEED_QFUNCTION_NONE,
-                       CEED_QFUNCTION_NONE, &op_prolong);
-    CeedOperatorSetField(op_prolong, "input", data[i-1]->elem_restr_u,
-                         basis_c_to_f, CEED_VECTOR_ACTIVE);
-    CeedOperatorSetField(op_prolong, "output", data[i]->elem_restr_u,
-                         CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
-
-    // Save libCEED data required for level
-    data[i]->op_prolong = op_prolong;
-  }
+  // Restriction - Fine to corse
+  CeedOperator op_restrict;
+  // Interpolation - Corse to fine
+  CeedOperator op_prolong;
+  // Coarse grid operator
+  CeedOperator op_apply;
+  // Basis
+  ierr = CreateBasisFromPlex(ceed, dm, 0, 0, 0, 0, &data[level-1]->basis_c_to_f);
```

Comment:
Note, here you wouldn't be providing the basis coarse-to-fine but rather the coarse grid basis

Reply:
I was going to ask you about that. I call `CeedLevelTransferSetup` in line 402 of multigrid.c 
```
if (i > 0) {
      // Prolongation/Restriction Operator
      ierr = CeedLevelTransferSetup(dm[i], ceed, i, num_levels, num_comp_u, ceed_data,
                                    level_degrees, mult[i]); CHKERRQ(ierr);
      user_pr[i]->comm = comm;
      user_pr[i]->dmf = dm[i];
      user_pr[i]->dmc = dm[i-1];
      user_pr[i]->loc_vec_c = X_loc[i-1];
      user_pr[i]->loc_vec_f = user_O[i]->Y_loc;
      user_pr[i]->mult_vec = mult[i];
      user_pr[i]->ceed_vec_c = user_O[i-1]->x_ceed;
      user_pr[i]->ceed_vec_f = user_O[i]->y_ceed;
      user_pr[i]->op_prolong = ceed_data[i]->op_prolong;
      user_pr[i]->op_restrict = ceed_data[i]->op_restrict;
      user_pr[i]->ceed = ceed;
    }
```
is it correct to pass the `dm` of fine level here?

Reply:
and I guess I don't need `num_levels` as an input anymore( I'll clean it up)

Reply:
I changed that to `basis_coarse`